File: environment/acceptance.rb
# frozen_string_literal: true
# Creates one disposable native-parity-* VM. No test framework gems required.
require_relative 'launcher'

name = ARGV.fetch(0)
raise 'Use a fresh native-parity-* name' unless name.match?(/\Anative-parity-[a-z0-9-]+\z/)
Dir.mktmpdir('native-parity-fixtures-') do |dir|
  dotfiles = File.join(dir, 'dotfiles')
  FileUtils.mkdir_p(dotfiles)
  File.write(File.join(dotfiles, 'install'), "#!/bin/bash\nset -eu\nprintf '%s\\n' \"$USER\" > \"$HOME/.parity-dotfiles\"\nprintf 'return 0\\n' > \"$HOME/.profile\"\n")
  File.chmod(0o755, File.join(dotfiles, 'install'))
  Environment.run('git', 'init', '-q', dotfiles)
  Environment.run('git', '-C', dotfiles, 'add', 'install')
  Environment.run('git', '-C', dotfiles, '-c', 'user.name=Test', '-c', 'user.email=test@example.invalid', 'commit', '-qm', 'Test installer')
  File.write(File.join(dir, 'system.sh'), "#!/bin/bash\nset -eu\necho once >> /usr/local/share/devwright/project-runs\n")
  File.write(File.join(dir, 'user.sh'), "#!/bin/bash\nset -eu\ntest \"$APP_ENV\" = 'literal {{braces}} $HOME'\necho keep > \"$HOME/projects/preserve\"\n")
  requirements = File.join(dir, 'requirements.toml')
  File.write(requirements, Environment.read('config/codex/requirements.toml') + "\n# Native parity policy\n")
  path = File.join(dir, 'config.yaml')
  File.write(path, YAML.dump({ 'packages' => ['tree'], 'dotfiles_repo' => dotfiles,
    'environment' => { 'APP_ENV' => 'literal {{braces}} $HOME' },
    'system_script' => 'system.sh', 'user_script' => 'user.sh', 'codex_requirements' => requirements,
    'credentials' => [{ 'name' => 'NATIVE_PARITY_TOKEN', 'description' => 'Synthetic {{test}} credential' }] }))
  # A missing credential fails unattended onboarding, after the guest is provisioned.
  out, err, status = Open3.capture3(File.join(__dir__, 'dev'), 'create', name, '--config', path)
  puts out
  warn err
  raise 'Expected missing-credential onboarding failure' if status.success? || !err.include?('Missing NATIVE_PARITY_TOKEN')
  state = Environment.info(name)
  values = { 'NATIVE_PARITY_TOKEN' => "first '$HOME' value" }
  Environment.capture(*Environment.ssh_args(state, 'dev', ['/bin/bash', '-s']), input: Environment.credential_script(values))
  values['NATIVE_PARITY_TOKEN'] = "second '$HOME' value"
  Environment.capture(*Environment.ssh_args(state, 'dev', ['/bin/bash', '-s']), input: Environment.credential_script(values))
  Environment.finish(state) # Checks existing credentials without a terminal.
  puts 'PASS fresh provisioning, missing-credential recovery, rotation, and finish'

  check = <<~'SH'
    set -euo pipefail
    test "$APP_ENV" = 'literal {{braces}} $HOME'
    test "$NATIVE_PARITY_TOKEN" = "second '\$HOME' value"
    test "$(cat "$HOME/.parity-dotfiles")" = dev
    test "$(cat "$HOME/projects/preserve")" = keep
    test -x /usr/bin/tree
    ! grep -q first "$HOME/.config/devwright/credentials.d/NATIVE_PARITY_TOKEN.sh"
    test "$(stat -c %a "$HOME/.config/devwright/credentials.d/NATIVE_PARITY_TOKEN.sh")" = 600
    grep -q 'Native parity policy' /etc/codex/requirements.toml
  SH
  Environment.run(*Environment.ssh_args(state, 'dev', ['/bin/bash', '-c', check]))
  Environment.run(*Environment.ssh_args(state, 'root', ['test', '-f', '/root/.parity-dotfiles']))
  root_check = 'test "$(wc -l < /usr/local/share/devwright/project-runs)" = 1'
  Environment.run(*Environment.ssh_args(state, 'root', ['/bin/bash', '-c', root_check]))
  completed = Environment.capture(*Environment.ssh_args(state, 'root', ['cat', '/usr/local/share/devwright/lima-setup-complete']))
  # Changing source files cannot affect the saved creation recipe or rerun setup.
  File.write(File.join(dir, 'system.sh'), "#!/bin/bash\nexit 99\n")
  File.write(path, YAML.dump({ 'packages' => ['not-a-real-package'] }))
  Environment.run('limactl', 'stop', name)
  Environment.run('limactl', 'start', '--tty=false', name)
  state = Environment.info(name)
  Environment.verify(state)
  Environment.run(*Environment.ssh_args(state, 'dev', ['/bin/bash', '-c', check]))
  Environment.run(*Environment.ssh_args(state, 'root', ['/bin/bash', '-c', root_check]))
  raise 'Setup reran on restart' unless completed == Environment.capture(*Environment.ssh_args(state, 'root', ['cat', '/usr/local/share/devwright/lima-setup-complete']))
  Environment.run('ssh', "lima-#{name}", 'test "$(id -un)" = dev')
  puts 'PASS restart skips setup, retains original recipe and user state, and keeps SSH alias working'
end

File: environment/acceptance.py
#!/usr/bin/env python3
"""Exercise preservation, overrides, dotfiles, credential transport, and restart.

Run only on a disposable VM created by environment/dev. This deliberately edits
that VM's personal settings and test credentials. It never reads real secrets.
"""
import argparse
from pathlib import Path
import tempfile
from unittest.mock import patch

import yaml
import launcher


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('name', help='Disposable VM; name must start with ansible-parity-')
    args = parser.parse_args()
    if not args.name.startswith('ansible-parity-'):
        parser.error('Use a disposable ansible-parity-* VM')
    state = launcher.info(args.name)
    with tempfile.TemporaryDirectory(prefix='ansible-parity-fixtures-') as tmp:
        directory = Path(tmp)
        dotfiles = directory / 'dotfiles'
        dotfiles.mkdir()
        installer = dotfiles / 'install'
        installer.write_text('#!/bin/bash\nset -eu\nprintf "%s\\n" "$USER" > "$HOME/.parity-dotfiles"\nprintf "return 0\\n" > "$HOME/.profile"\n')
        installer.chmod(0o755)
        launcher.run(['git', 'init', '-q', dotfiles])
        launcher.run(['git', '-C', dotfiles, 'add', 'install'])
        launcher.run(['git', '-C', dotfiles, '-c', 'user.name=Parity Test', '-c', 'user.email=parity@example.invalid',
                      'commit', '-qm', 'Test dotfiles installer'])
        requirements = directory / 'requirements.toml'
        requirements.write_text((launcher.ROOT / 'config/codex/requirements.toml').read_text() + '\n# custom parity policy\n')
        managed = directory / 'managed.json'
        managed.write_text((launcher.ROOT / 'config/claude/managed-settings.json').read_text())
        codex = directory / 'codex.toml'
        codex.write_text((launcher.ROOT / 'config/codex/config.toml').read_text() + '\n# explicit parity replacement\n')
        claude = directory / 'claude.json'
        claude.write_text((launcher.ROOT / 'config/claude/settings.json').read_text())
        recipe = {
            'dotfiles_repo': str(dotfiles), 'packages': ['tree'],
            'environment': {'PARITY_ENV': "literal $HOME and 'quotes'"},
            'codex_requirements': str(requirements), 'claude_managed_settings': str(managed),
            'codex_config': str(codex), 'claude_config': str(claude),
        }
        config_file = directory / 'recipe.yaml'
        launcher.ssh(state, 'dev', ['python3', '-c', '''from pathlib import Path
import json
p=Path.home()/'.codex/config.toml'; p.write_text(p.read_text()+'\\n# preserve-parity-personal\\n')
p=Path.home()/'.claude/settings.json'; d=json.loads(p.read_text()); d['env']={'PARITY_PRESERVE':'yes'}; p.write_text(json.dumps(d))
(Path.home()/'projects/parity-preserve').write_text('keep me')
'''])
        declarations = [{'name': 'PARITY_CREDENTIAL', 'description': 'Synthetic test value'}]
        for value in ["first '$HOME' value", "second '$HOME' value"]:
            with patch.object(launcher.sys.stdin, 'isatty', return_value=True), patch.object(launcher.getpass, 'getpass', return_value=value):
                launcher.credentials(state, declarations, replace=True)

        def apply():
            config_file.write_text(yaml.safe_dump(recipe))
            launcher.run([launcher.HERE / 'dev', 'configure', args.name, '--config', config_file])

        def assert_guest(preserved, custom):
            code = '''import json,os
from pathlib import Path
h=Path.home()
assert (h/'projects/parity-preserve').read_text()=='keep me'
assert os.environ['PARITY_ENV']=="literal $HOME and 'quotes'"
assert os.environ['PARITY_CREDENTIAL']=="second '$HOME' value"
assert "first " not in (h/'.config/devwright/credentials.sh').read_text()
assert (h/'.parity-dotfiles').read_text().strip()=='dev'
assert os.access('/usr/bin/tree',os.X_OK)
'''
            if preserved:
                code += "assert 'preserve-parity-personal' in (h/'.codex/config.toml').read_text()\n"
                code += "assert json.loads((h/'.claude/settings.json').read_text())['env']['PARITY_PRESERVE']=='yes'\n"
            else:
                code += "assert 'explicit parity replacement' in (h/'.codex/config.toml').read_text()\n"
                code += "assert 'preserve-parity-personal' not in (h/'.codex/config.toml').read_text()\n"
                code += "assert 'env' not in json.loads((h/'.claude/settings.json').read_text())\n"
            launcher.ssh(state, 'dev', ['python3', '-c', code])
            root_code = f'''from pathlib import Path
p=Path('/usr/local/share/devwright')
assert (p/'custom-requirements.toml').exists()=={custom!r}
assert (p/'custom-managed-settings.json').exists()=={custom!r}
assert Path('/root/.parity-dotfiles').read_text().strip()=='root'
assert ('custom parity policy' in Path('/etc/codex/requirements.toml').read_text())=={custom!r}
'''
            launcher.ssh(state, 'root', ['python3', '-c', root_code])

        apply()
        assert_guest(preserved=True, custom=True)
        print('PASS custom policy selection, default preservation, host-bundled dotfiles, environment, credential rotation', flush=True)
        for key in ('codex_requirements', 'claude_managed_settings', 'codex_config', 'claude_config'):
            recipe.pop(key)
        apply()
        assert_guest(preserved=True, custom=True)
        print('PASS repeated configuration preserves selected policy and user state', flush=True)
        recipe |= {'reset_codex_requirements': True, 'reset_claude_managed_settings': True,
                   'codex_config': str(codex), 'claude_config': str(claude),
                   'replace_codex_config': True, 'replace_claude_config': True}
        apply()
        assert_guest(preserved=False, custom=False)
        print('PASS explicit replacement and managed-policy reset', flush=True)
        launcher.run(['limactl', 'stop', args.name])
        launcher.run(['limactl', 'start', '--tty=false', args.name])
        state = launcher.info(args.name)
        launcher.verify(state)
        assert_guest(preserved=False, custom=False)
        print('PASS restart retains isolation and SSH connectivity', flush=True)


if __name__ == '__main__':
    main()
